ob-jq.el

```diff
@@ -49,52 +50,85 @@
-		      (with-temp-file file
-			(insert body)
-			file)))
-	 (stdin (let ((stdin (cdr (assq :stdin params))))
-		  (when stdin
-		    (let ((tmp (org-babel-temp-file "jq-stdin-"))
-			  (res (org-babel-ref-resolve stdin)))
-		      (with-temp-file tmp
-			(insert res)
-			tmp)))))
-	 (cmd (mapconcat #'identity
-			 (remq nil
-			       (list org-babel-jq-command
-				     (format "--from-file \"%s\"" code-file)
-				     cmd-line
-				     in-file))
-			 " ")))
+         (cmd-line (cdr (assq :cmd-line params)))
+         (in-file (cdr (assq :in-file params)))
+         (code-file (let ((file (org-babel-temp-file "jq-")))
+                      (with-temp-file file
+                        (insert body)
+                        file)))
+         (stdin (let ((stdin (cdr (assq :stdin params))))
+                  (when stdin
+                    (let ((tmp (org-babel-temp-file "jq-stdin-"))
+                          (res (org-babel-ref-resolve stdin)))
+                      (with-temp-file tmp
+                        (insert
+                         (cond
+                          ((stringp res) res)
+                          ((listp res) (org-babel-jq-table-to-json res))))
+                        tmp)))))
+         (cmd (mapconcat #'identity
+                         (remq nil
+                               (list org-babel-jq-command
+                                     (format "--from-file \"%s\"" code-file)
+                                     cmd-line
+                                     in-file))
+                         " ")))
     (org-babel-reassemble-table
      (let ((results
-	    (cond
-	     (stdin (with-temp-buffer
-		      (call-process-shell-command cmd stdin (current-buffer))
-		      (buffer-string)))
-	     (t (org-babel-eval cmd "")))))
+            (cond
+             (stdin (with-temp-buffer
+                      (call-process-shell-command cmd stdin (current-buffer))
+                      (buffer-string)))
+             (t (org-babel-eval cmd "")))))
        (when results
-	 (org-babel-result-cond result-params
-	   results
-	   (let ((tmp (org-babel-temp-file "jq-results-")))
-	     (with-temp-file tmp
-	       (insert results))
-	     (org-babel-import-elisp-from-file tmp)))))
+         (org-babel-result-cond result-params
+           results
+           (let ((data (json-read-from-string results)))
+             ;; If we have an array we might have a table
+             (if (and (vectorp data)
+                      (> (length data) 0))
+                 (cond
+                  ;; If the first element is a vector then just "unpack"
+                  ;; the vector of vectors
+                  ((vectorp (aref data 0))
+                   (mapcar (lambda (row) (append row nil)) data))
+                  ;; If the first element is a list we will assume we
+                  ;; have an array of objects, so generate the colnames
+                  ;; accordingly
```

Comment:
Is this an OK assumption? What happens if we get a list of different objects? F.ex. `[{"a" : 1}, {"b" : 2 }]`. 

I'm sorry to give you more work. I'm just not very familiar with the json-package and how it works so I can't immediately see what is going on here. If you don't have time I will check it myself later. I just need to be convinced it won't blow up in some horrible way when it gets some strange result data.

Reply:
I think the `json` package is not relevant here.  The assumption is on the side of the user.  If we don't specify that we want a table result the output is always returned as "json string".  If the user specifies `:results table` then we assume the data is in a "table-compatible" way.

We can of course add an assertion and then default to some other type of output but maybe it's even better to error out and let the user know that "Hey, you want a table but we can't do that because the output is not a table, you probably have an error somewhere".

I can add a check that the returned vector is homogenous, that should be fairly simple.

Reply:
Yes, of course. I am sorry. I was quite tired yesterday. I was just worried that this would happen to all results and then it would be difficult to debug. I should have thought that it of course is only valid for tables. That is not a big problem. I'm not entirely sure we need a check, just a graceful error. I would assume the org function that writes the table will error for us if something is up.